src/operator/contrib/csr_neighborhood_sample-inl.h

```diff
@@ -0,0 +1,281 @@
+                      std::vector<dgl_id_t>& out) {
+  // Copy ver_list to output
+  if (ver_list.size() <= max_num_neighbor) {
+    for (size_t i = 0; i < ver_list.size(); ++i) {
+      out.push_back(ver_list[i]);
+    }
+    return;
+  }
+  // Make sample
+  std::unordered_map<dgl_id_t, bool> mp;
+  while (out.size() < max_num_neighbor) {
+    random_shuffle(ver_list.begin(), ver_list.end());
+    for (size_t i = 0 ; i < ver_list.size(); ++i) {
+      int rand_num = (rand() % 100) + 1; 
+      float prob = probility[ver_list[i]];
+      if (rand_num <= static_cast<int>(prob*100) &&
+          mp[ver_list[i]] == false) {
+        out.push_back(ver_list[i]);
+        mp[ver_list[i]] = true;
+        if (out.size() >= max_num_neighbor) {
+          return;
+        }
+      }
+    }
+  }
+}
+
+static void CSRNeighborSampleComputeExCPU(const nnvm::NodeAttrs& attrs,
+                                          const OpContext& ctx,
+                                          const std::vector<NDArray>& inputs,
+                                          const std::vector<OpReqType>& req,
+                                          const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+
+  const NeighborSampleParam& params = 
+    nnvm::get<NeighborSampleParam>(attrs.parsed);
+
+  // set seed for random sampling
+  srand(time(nullptr));
+
+  dgl_id_t num_hops = params.num_hops;
+  dgl_id_t num_neighbor = params.num_neighbor;
+  dgl_id_t max_num_vertices = params.max_num_vertices;
+
+  size_t vertices_num = inputs[1].data().Size();
+  size_t egde_num = inputs[0].data().Size();
+  size_t seed_num = inputs[2].data().Size();
+
+  CHECK_GE(max_num_vertices, seed_num);
+
+  const dgl_id_t* val_list = inputs[0].data().dptr<dgl_id_t>();
+  const dgl_id_t* col_list = inputs[0].aux_data(1).dptr<dgl_id_t>();
+  const dgl_id_t* indptr = inputs[0].aux_data(0).dptr<dgl_id_t>();
+  const dgl_id_t* seed = inputs[2].data().dptr<dgl_id_t>();
+  const float* prob_array = inputs[1].data().dptr<float>();
+
+  size_t row_len = inputs[0].aux_data(0).Size() - 1;
+  CHECK_EQ(row_len, vertices_num);
+
+  dgl_id_t* out = outputs[0].data().dptr<dgl_id_t>();
+
+  // Get the mapping between edge_id and row_id
+  dgl_id_t idx = 0;
+  std::unordered_map<dgl_id_t, dgl_id_t> edge_mp;
+  for (size_t i = 1; i < vertices_num+1; ++i) {
+    size_t edge_in_a_row = indptr[i] - indptr[i-1];
+    for (size_t j = 0; j < edge_in_a_row; ++j) {
+      edge_mp[val_list[idx++]] = i-1;
+    }
+  }
+
+  // Get the mapping between col_id (src) and row_id_list (dst)
+  std::vector<std::vector<dgl_id_t> > col_row_list(vertices_num);
+  for (size_t i = 0; i < egde_num; ++i) {
+    dgl_id_t row_id = edge_mp[val_list[i]];
+    dgl_id_t col_id = col_list[i];
+    col_row_list[col_id].push_back(row_id);
+  }
+
+  // BFS traverse the graph and sample vertices
+  std::vector<dgl_id_t> sub_ver_mp(vertices_num, 0);
```

Comment:
it's not a good idea to allocate a vector for all nodes in the input graph. A graph can be super large.

Reply:
Yes. Here we can just use the unordered_map to store the flag. Here I want to speedup the memory access but the memory cost could be high.